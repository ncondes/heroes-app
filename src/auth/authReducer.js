import { types } from '../types/types';

// const state = {
//     name: 'Nicolas',
//     logged: true
// }

// const loginAction = {
//     type: types.login,
//     payload: {
//         name: 'Nicolas',
//         email: 'Nicolas@gmail.com'
//     }
// }

export const authReducer = ( state = {}, action ) => {

    switch ( action.type ) {
        case types.login:
            return {
                ...action.payload,
                logged: true,
            }
        case types.logout:
            return {
                logged: false,
            }
    
        default:
            return state;
    }

}
