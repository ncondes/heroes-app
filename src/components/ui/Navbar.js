import { useContext } from 'react';
import { Link, NavLink, useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { types } from '../../types/types';

export const Navbar = () => {

    const { user, dispatch } = useContext( AuthContext );
    // console.log( user );

    const navigate = useNavigate();

    const handleLogout = () => {

        const action = {
            type: types.logout,
            payload: {},
        }

        dispatch( action );

        navigate('login', {
            replace: true,
        })
    }

    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
            <div className='container-fluid'>

            <Link 
                className="navbar-brand" 
                to="/"
            >
                Asociaciones
            </Link>

            <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarNavAltMarkup" aria-controls="navbarNavAltMarkup" aria-expanded="false" aria-label="Toggle navigation">
                <span className="navbar-toggler-icon"></span>
            </button>
        
                <div className='navbar-collapse collapse justify-content-between' id='navbarNavAltMarkup'>


                    <div className="navbar-nav">
                        <NavLink 
                            className={ ({ isActive }) => `nav-link ${ isActive && 'active' }`}
                            aria-current='page'
                            to="/marvel"
                        >
                            Marvel
                        </NavLink>

                        <NavLink 
                            className={ ({ isActive }) => `nav-link ${ isActive && 'active' }`} 
                            to="/dc"
                        >
                            DC
                        </NavLink>

                        <NavLink 
                            className={ ({ isActive }) => `nav-link ${ isActive && 'active' }`} 
                            to="/search"
                        >
                            Search
                        </NavLink>
                    </div>

                    <div className="navbar-nav">
                        <span
                            className='nav-link text-info'
                        >
                            { user.name }
                        </span>

                        <button 
                            className="nav-link btn"
                            onClick={ handleLogout }
                        >
                            Logout
                        </button>
                    </div>
                </div>
            </div>
        </nav>
    )
}