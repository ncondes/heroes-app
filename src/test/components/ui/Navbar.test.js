import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import { Navbar } from '../../../components/ui/Navbar';
import { AuthContext } from '../../../auth/authContext';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('Pruebas en Navbar', () => {

    test('Debe de mostrarse correctamente', () => {

        const contextValue = {
            dispatch: jest.fn(),
            user: {
                name: 'pedro',
                logged: true,
            },
        }

        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter initialEntries={ ['/'] }>
                    <Navbar />
                </MemoryRouter>
            </AuthContext.Provider>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.text-info').text().trim() ).toBe( 'pedro' )

    });

    test('Debe de llamar el logout, llamar el navigate y el dispatch con los argumentos', () => {
        
        const contextValue = {
            dispatch: jest.fn(),
            user: {
                name: 'pedro',
                logged: true,
            },
        }

        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter initialEntries={ ['/'] }>
                    <Navbar />
                </MemoryRouter>
            </AuthContext.Provider>
        );

        wrapper.find('button').at( 1 ).prop('onClick')();

        expect( contextValue.dispatch ).toHaveBeenCalledWith({"payload": {}, "type": "[auth] Logout"});
        expect( mockNavigate ).toHaveBeenCalledWith("login", {"replace": true});

    })

})