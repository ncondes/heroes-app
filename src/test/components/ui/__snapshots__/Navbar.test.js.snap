// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Navbar Debe de mostrarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <div
    className="container-fluid"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
    <button
      aria-controls="navbarNavAltMarkup"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-bs-target="#navbarNavAltMarkup"
      data-bs-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="navbar-collapse collapse justify-content-between"
      id="navbarNavAltMarkup"
    >
      <div
        className="navbar-nav"
      >
        <a
          className="nav-link false"
          href="/marvel"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          className="nav-link false"
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          className="nav-link false"
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
      <div
        className="navbar-nav"
      >
        <span
          className="nav-link text-info"
        >
          pedro
        </span>
        <button
          className="nav-link btn"
          onClick={[Function]}
        >
          Logout
        </button>
      </div>
    </div>
  </div>
</nav>
`;
